### ISSUE TITLE
Feature request restart on Windows

### ISSUE THREAD

**Issue by SteffenAL** [icing/mod_md#17]
In the Readme:
When mod_md gets a certificate from Let's Encrypt, installs it in your file system and restart the server to activate it.

For Windows  to stop/restart auto the server must be an option, best is to sent an email that the certs are renewed or it can start a script.  On windows stopping/restarting mostly leaves e.g. mod_fcgid processes running. When I restart Apache, I run a script that also kills all the mod_fcgid processes.

**Comment by Jan-E** [comment]
@SteffenAL Why would you want to kill the php-cgi.exe's when a certificate is installed or renewed? AFAIK nothing changes for PHP. It is Apache that is serving the certificates, not PHP.

Or will the php-cgi.exe instances be stale in the memory, doing nothing because their previous mod_fcgid.so process does not exist anymore? Would that be a real problem until you manually restart the server in due time?

**Comment by SteffenAL** [comment]
It is a well-known issue on Windows that when you stop/restart apache that not always php-cgi.exe processes are stopped/killed. So after a restart new php-cgi processes are there, and the (zombies) processes before the stop/restart are still there doing nothing.

**Comment by Jan-E** [comment]
Which situation is worse:
1. Running Apache with a certificate in memory that has been replaced on disk
2. Having zombie php-cgi.exe's

I would not be too sure about Apache with different certificates.

I haven't seen the zombie problem lately and could not reproduce it in a quick test. But the problem is solvable with sysinternal tools.

pslist -t shows something like;

```
C:\>pslist -t

PsList v1.4 - Process information lister
Copyright (C) 2000-2016 Mark Russinovich
Sysinternals - www.sysinternals.com

Process information:

Name                             Pid Pri Thd  Hnd      VM      WS    Priv
....
    httpd                       3768   8   8  324  203524   32876   33932
      httpd                     3904   8 387 1136  430660  107364   70620
        php-cgi                 8360   8   5  218  339048   61364   23836
        php-cgi                 8368   8   4  107  328868   39400   22160
    httpd                       5388   8   8  285  201936   31852   33600
      httpd                     4676   8 390 1121  435860   69480   81780
    httpd                       6292   8   8  288  201936   31992   33728
      httpd                     3052   8 387 1147  479332  132544  115112
        php-cgi                 8396   8   5  117  394372   96128   69096
        php-cgi                10828   8   5  113  394372   99136   70204
```
Quiet server with 3 of my 4 Apaches in the list.
A zombie php-cgi.exe would have no pair of httpd parents and show up like

```
Name                             Pid Pri Thd  Hnd      VM      WS    Priv

php-cgi                        10828   8   5  113  394372   99136   70204
```

Parse the output of 'pslist -t' in your favourite parser (mine would be php), find every php-cgi without parents and pskill those processes, using the pid in the second column. Make it a taskmanager script and run it every day, every hour or whatever. No more zombies.

**Comment by icing** [comment]
@Jan-E Do you have code recommendations for a service restart function? I have a place in ```mod_md/md_os.c``` for OS specific functions. I see the code in ```mpm_winnt.c``` but since I do not compile myself under Windows, maybe one of you could give it a try?

**Comment by SteffenAL** [comment]
In Apachemonitor.c is also a service restart function.

Recap:

My proposal is:
MDRestartWin = Auto/No/script

For example a script can sent an email,  kill zombie processes, copy  e.g. cert's to mail servers etc. I use that kind of script with my current  ACME tool (letsencrypt-win-simple).

**Comment by Jan-E** [comment]
TL:DR I do not know if there is a way to restart the current instance of Apache.

> In Apachemonitor.c is also a service restart function.

I was also looking at the code there. Obstacle: how do you know which Apache service should be restarted?

ApacheMonitor collects a list of all services in services.msc (no less than 9 on my laptop). You can start/stop all of them.

The same accounts for net start/stop:
```
D:\>net stop Apache24.x64
The Apache24.x64 service is not started.
```
And httpd.exe also needs a service name. You can start/stop other Apache instances, even a x64 Apache from a x86 Apache:
```
D:\>\phpdev\apache24.x32\bin\httpd.exe -n Apache24.x64 -k stop
The 'Apache24.x64' service is not started.
```
The service name defaults to Apache2.4:
```
D:\>\phpdev\apache24.x64\bin\httpd.exe -k restart
[Thu Aug 03 18:10:23.691933 2017] [mpm_winnt:error] [pid 8188:tid 120] (OS 2)
The system cannot find the file specified.  : AH00436: No installed service named "Apache2.4".
```
Looking from outside to a running Apache there seems te be no way to know what the service name is.
'net start' lists the running services, but if you have got more than one, you get something like:
```
C:\>net start
These Windows services are started:

   Apache007Portal
   Apache226Pma
   Apache227Fimp
   Apache228Sessie
   Application Experience
  ...
```

**Comment by Jan-E** [comment]
Stackoverflow;
How can a Windows Service determine its ServiceName?

Sort of solution:
https://stackoverflow.com/a/1877404/872051

If there is no internal command in Apache to restart it self, you''l need clumsy workarounds.

**Comment by JBlond** [comment]
@SteffenAL wouldnn't it be better to fix mod_fcgid that 
There is a ticket for that https://bz.apache.org/bugzilla/show_bug.cgi?id=48769 and https://bugs.php.net/bug.php?id=54293
I had a lot of mails about that topic with Jeff Trawick from 2013 - 2015 but no solving that issue.

**Comment by icing** [comment]
@Jan-E Thanks for digging this  up. No, there currently is no internal "restart gracefully" service at the moment. And yes, I think there should be.

But I see that the solution of this across platforms is not trivial. So, I tend to think this needs to be done in the core part of the server, if at all. Also the suggestion by @SteffenAL about starting a script would fit more into such an approach.

My vision would be something like ```mod_ctrl```, offering an API like:

```
   typedef enum {
     AP_CTRL_RESTART_WHEN_CONVENIENT,
     AP_CTRL_RESTART_NOW,
   } ap_ctrl_restart_type_t;
   apr_status_t ap_ctrl_restart(ap_ctrl_restart_type_t type, const char *reason);
```
and directives like:

```
  RestartWindow  02:00h-03:00h
  RestartWindow  00:00h-06:00h sundays

  # never = do not do it yourself, but run notifications
  Restart never | auto
  # run a script when restart is requested or performed
  RestartRequestNotification <path to exec given type+reason as arg>
  RestartNotification <path to exec given type+reason as arg>
```

Some volunteers to write this?

**Comment by SteffenAL** [comment]
Closing:  the feature request is heard by author.

**Comment by SteffenAL** [comment]
Just to remind the request starting a script, like @icing suggested:

  # never = do not do it yourself, but run notifications
  Restart never | auto
  # run a script when restart is requested or performed
  RestartRequestNotification <path to exec given type+reason as arg>
  RestartNotification <path to exec given type+reason as arg>

**Comment by icing** [comment]
@SteffenAL leave this open please.

**Comment by SteffenAL** [comment]
I came to an simple solution, no directive needed.

Changed  mod_md_os.c  to :
#ifdef WIN32
apr_status_t md_server_graceful(apr_pool_t *p, server_rec *s)
{
  system ("mod_md.bat \"Certificate(s) renewed\" \"See error.log for more info\" \"mod_md.bat is excuted\" ");
 return 0 ;
}

In the **/Windows folder**  I have **mod_md.bat** and **mailsend.exe**

In mod_md.bat you can put all the things/commands you want, no need for directive as mentioned above.  **Example:**

@ECHO OFF
REM Remove following  if you do not want to restart
REM You can also do:  net start service-name followed by net start service-name
httpd -k restart 

mailsend -q -f name@abc.com -smtp  xyz.com -user name@def.coml -pass xxxxx  -name "Steffen L" -t name@ghi.com -sub "Lets Encrypt mod_md notification" -M "%~1" -M "%~2"  -M "%~3" -M "%~4" -M "."
 
**note:**  the arguments are passed to : -M "%~1" -M "%~2"  -M "%~3" -M "%~4" -M "."

This give me a nice mail with:

_Subject:  Lets Encrypt mod_md notification
From: Steffen L
Certificate(s) renewed
See error.log for more info
mod_md.bat is excuted_



**Note:**
mailsend.exe , see https://github.com/muquit/mailsend/releases and https://github.com/muquit/mailsend/blob/master/doc/examples.mediawiki

**Comment by icing** [comment]
@SteffenAL please try v1.0.0 and the new config directive ```MDNotifyCmd``` if that works well on your system also.

Although you approach works, I added the config directive for security reasons. The server should not run scripts blindly from locations where other users might be able to install their code.

**Comment by SteffenAL** [comment]
The ZIP is still 0.9.9

**Comment by icing** [comment]
The src/md_version.h itself is not in the tar. Is that what you mean?

**Comment by icing** [comment]
Ah, my fault. Did not push - the common git mistake. Correct version is now in github.

**Comment by SteffenAL** [comment]
Removed the /md folder and  certificate is created.
Looks ok, I got a nice mail with domain name and stop/start.

Attention:
When back slashes in path MDNotifyCmd **"**D:\servers\apache\bin\mod_md.bat**"**  then I get:
**'"**D:\servers\apache\serversapachebinmod_md.bat**"'** is not recognized as an internal or external command, operable program or batch file.

Ok when forward slashes MDNotifyCmd D:/servers/apacheS/bin/mod_md.bat
mod_md.c(764): AH: notify command 'D:/servers/apacheS/bin/mod_md.bat' returned 0

**Comment by icing** [comment]
Ok, should add that to the docs. But otherwise, this works as desired?

**Comment by SteffenAL** [comment]
See also http://httpd.apache.org/docs/trunk/platform/windows.html

• The directives that accept filenames as arguments must use Windows filenames instead of Unix ones. However, because Apache may interpret backslashes as an "escape character" sequence, you should consistently use forward slashes in path names, not backslashes.

But there are Directives which accept back slashes on windows, for example:

PidFile D:\servers\apacheS\httpd.pid
ServerRoot "D:\servers\apacheS"

**Comment by icing** [comment]
Ok, probably depends on the library calls such paths get used in. Windows admins should be safe with using / in configs.

**Comment by icing** [comment]
Hey, @oerdnj and @Sp1l - kind souls that you are!

I would like to propose a backport of ```mod_md``` into the 2.4.x branch pretty soon and if you two could once more crank out a release on the current v1.0.0 that'd be great! We need all the tests we can get...and then some. :-D

**Comment by SteffenAL** [comment]
Maybe we should  Notify when some is going bad/error to create/renew.

**Comment by icing** [comment]
This is a slippery slope. I think error log monitoring is a good idea in general. Having an error notification channel for each module seems not the way forward.

The current MDNotifyCmd is a crutch until there is a module for restarts and related things. I am certain mod_md is not the last module to make use of such a facility.

**Comment by Sp1l** [comment]
Your wish was granted!
https://svnweb.freebsd.org/ports?view=revision&sortby=date&revision=451851

Would love to see this appear in next Apache 2.4 :D

**Comment by icing** [comment]
❤️!

**Comment by SteffenAL** [comment]
Better to code this.
